rfcs/2022-04-20-12187-log-namespacing.md

```diff
@@ -0,0 +1,422 @@
+  "stream": "stderr",
+  "source_type": "kubernetes_logs",
+  "ingest_timestamp": "2020-10-15T11:01:46.499555308Z"
+}
+```
+
+-----
+
+
+#### Syslog / Vector namespace
+
+event (the "data" key was elided)
+
+```json
+{
+    "message": "Hello Vector",
+    "hostname": "localhost",
+    "severity": "info",
+    "facility": "facility",
+    "appname": "Vector Hello World",
+    "msgid": "238467-435-235-a3478fh",
+    "procid": 13512,
+    // this name is up for debate. Arbitrary keys need to be nested under something though
+    "structured_data": {
+      "origin": "timber.io"
+    }
+}
+```
+
+metadata
+
+```json
+{
+  "log_namespace": "vector",
+  "source_ip": "127.0.0.1",
+  "hostname": "localhost",
+  "source_type": "syslog",
+  "ingest_timestamp": "2020-10-15T11:01:46.499555308Z"
+}
+```
+
+-----
+
+#### Socket source (mode=udp) / syslog codec / Vector namespace
+
+event (the "data" key was elided)
+
+```json
+{
+    "message": "Hello Vector",
+    "hostname": "localhost",
+    "severity": "info",
+    "facility": "facility",
+    "appname": "Vector Hello World",
+    "msgid": "238467-435-235-a3478fh",
+    "procid": 13512
+}
+```
+
+metadata
+
+```json
+{
+  "log_namespace": "vector",
+  "source_ip": "192.168.0.1",
+  "hostname": "localhost",
+  "source_type": "socket",
+  "ingest_timestamp": "2020-10-15T11:01:46.499555308Z"
+}
+```
+
+-----
+
+#### HTTP source / JSON codec / Vector namespace
+
+event
+
+```json
+{
+  "mineral": "quartz",
+  "food": "sushi"
+}
+```
+
+metadata
+
+```json
+{
+  "path": "/foo/bar",
+  // headers and query params were previously placed directly on the root. This needs to be nested to avoid potential naming conflicts.
+  "headers": {
+    "Content-Type": "application/json"
+  },
+  "query_params": {
+    "page": 14,
+    "size": 3
+  }
+  "log_namespace": "vector",
+  "source_type": "http",
+  "ingest_timestamp": "2020-10-15T11:01:46.499555308Z"
+}
+```
+
+-----
+
+#### Kafka source / Native codec / Vector namespace
+
+This is an example where an event came from a Kafka source (JSON codec), through a Kafka sink (using native codec), and then back out a Kafka source (native codec).
+Notice that since `key` and `headers` wasn't moved into the event (from the event metadata), those values from the first kafka source were lost.
+
+event
+
+```json
+{
+  "derivative": -2.266778047142367e+125,
+  "integral": "13028769352377685187",
+  "mineral": "H 9 ",
+  "proportional": 3673342615,
+  "vegetable": -30083
+}
+```
+
+metadata (only from the 2nd kafka source)
+
+```json
+{
+  "key": "the key of the message (from the 2nd kafka source)"
+  "headers": {
+    "header-a-key": "header-a-value (from the 2nd kafka source)",
+    "header-b-key": "header-b-value (from the 2nd kafka source)"
+  }
+  "log_namespace": "vector",
+  "topic": "name of topic",
+  "partition": 3,
+  "offset": 1829448,
+  "source_type": "kafka",
+  "ingest_timestamp": "2022-04-14T19:14:21.899623781Z"
+}
+```
+
+## Outstanding Questions / Alternatives
```

Comment:
I think we'll need to ensure that this new metadata _does_ pass through the disk buffer to be usable in sink templating.

Reply:
Long term we probably need to. Initially the plan was to require it to be copied into the event to be used in a sink. There is no syntax yet to refer to metadata, and as you pointed out it doesn't exist once it goes through a disk buffer. We need to solve our initial concerns with storing this data to disk (since there are plain text api keys in the metadata).